tests/test_checkers.py

```diff
@@ -1,261 +1,357 @@
+        for v in values:
 
-    check_tester.run_a_checker(valids, invalids, CheckFloat, item='time_out')
+            config[section][item] = v
+            b = checker(config=user_config, section=section, item=item)
+            msgs = b.check()
 
+            if len([True for m in msgs if m is None]) == len(msgs):
+                valid = True
+            else:
+                valid = False
 
-def test_int(check_tester):
-    """
-    Test we see int as ints and not floats
-    """
+            # Expected valid
+            if z == 0:
+                assert valid
+            else:
+                assert not valid
 
-    # Confirm we these values are valid
-    valids = [10, '2', 1.0]
-    invalids = ['tough', '1.5', '']
-    check_tester.run_a_checker(valids, invalids, CheckInt, item='num_users')
 
+def define_checker(user_config, checker, value, item='item', section='basic'):
+    config = user_config.cfg
+    config.update({section: {item: " "}})
+    config[section][item] = value
 
-def test_datetime(check_tester):
-    """
-    Test we see datetime as datetime
-    """
+    return checker(config=user_config, section=section, item=item)
 
-    valids = ['2018-01-10 10:10', '10-10-2018', "October 10 2018"]
-    invalids = ['Not-a-date', 'Wednesday 5th']
-    check_tester.run_a_checker(
-        valids,
-        invalids,
-        CheckDatetime,
-        item='start_date')
 
+class TestCheckers:
+    @pytest.fixture(scope='class')
+    def master_config(self, test_config_dir):
+        """
+        Master config from test data
+        """
+        return MasterConfig(
+            path=test_config_dir.joinpath('master.ini').as_posix()
+        )
+
+    @pytest.fixture
+    def user_config(self, test_config_dir, master_config):
+        return UserConfig(
+            test_config_dir.joinpath('base_cfg.ini').as_posix(),
+            mcfg=master_config
+        )
+
+    def test_string(self, user_config):
+        """
+        Test we see strings as strings
+        """
+        valid_entries = ['test']
+
+        run_a_checker(
+            user_config,
+            valid_entries,
+            [],
+            checkers.CheckString,
+            item='username'
+        )
+
+        # Confirm that casting a string with uppers will auto produce lowers
+        user_config.cfg['basic']['username'] = 'Test'
+
+        b = checkers.CheckString(
+            config=user_config, section='basic', item='username'
+        )
+        result = b.cast()
+        assert result == 'test'
+
+        # Check we can capture a single item list for strings
+        b.is_list = True
+        result = b.cast()
+        assert result == ['test']
+
+        # Check we capture the when a list is passed and were not expecting one
+        b.is_list = False
+        result = b.cast()
+        assert not isinstance(result, list)
+
+    @pytest.mark.parametrize(
+        'value', [True, False, 'true', 'FALSE', 'yes', 'y', 'no', 'n']
+    )
```

Comment:
This recipe could be applied to all the below cases, where you combined them with the helper method. Using the pytest `parameterize` features makes them more granular and on failure exactly which of the ones in failed.

Reply:
Thanks for this! Super helpful.  I think I am still missing some understanding of the fixtures but this will get me moving forward.